sdk/js/src/utils/array.ts

```diff
@@ -1,72 +1,97 @@
-  CHAIN_ID_OASIS,
-  CHAIN_ID_POLYGON,
+  CHAIN_ID_ALGORAND,
   CHAIN_ID_SOLANA,
   CHAIN_ID_TERRA,
+  isEVMChain,
 } from "./consts";
 
-export const isEVMChain = (chainId: ChainId) => {
-  return (
-    chainId === CHAIN_ID_ETH ||
-    chainId === CHAIN_ID_BSC ||
-    chainId === CHAIN_ID_ETHEREUM_ROPSTEN ||
-    chainId === CHAIN_ID_AVAX ||
-    chainId === CHAIN_ID_POLYGON ||
-    chainId === CHAIN_ID_OASIS ||
-    chainId === CHAIN_ID_AURORA ||
-    chainId === CHAIN_ID_FANTOM ||
-    chainId === CHAIN_ID_KARURA ||
-    chainId === CHAIN_ID_ACALA ||
-    chainId === CHAIN_ID_KLAYTN
-  );
-};
+export { isEVMChain };
+
+/**
+ *
+ * Returns true iff the hex string represents a native Terra denom.
+ *
+ * Native assets on terra don't have an associated smart contract address, just
+ * like eth isn't an ERC-20 contract on Ethereum.
+ *
+ * The difference is that the EVM implementations of Portal don't support eth
+ * directly, and instead require swapping to an ERC-20 wrapped eth (WETH)
+ * contract first.
+ *
+ * The Terra implementation instead supports Terra-native denoms without
+ * wrapping to CW-20 token first. As these denoms don't have an address, they
+ * are encoded in the Portal payloads by the setting the first byte to 1.  This
+ * encoding is safe, because the first 12 bytes of the 32-byte wormhole address
+ * space are not used on Terra otherwise, as cosmos addresses are 20 bytes wide.
+ */
+export const isHexNativeTerra = (h: string): boolean => h.startsWith("01");
 
-export const isHexNativeTerra = (h: string) => h.startsWith("01");
-export const nativeTerraHexToDenom = (h: string) =>
+export const nativeTerraHexToDenom = (h: string): string =>
   Buffer.from(stripZeros(hexToUint8Array(h.substr(2)))).toString("ascii");
-export const uint8ArrayToHex = (a: Uint8Array) =>
+
+export const nativeTerraUint8ArrayToDenom = (a: Uint8Array): string =>
+  Buffer.from(stripZeros(a)).toString("ascii");
+
+export const uint8ArrayToHex = (a: Uint8Array): string =>
   Buffer.from(a).toString("hex");
-export const hexToUint8Array = (h: string) =>
+
+export const hexToUint8Array = (h: string): Uint8Array =>
   new Uint8Array(Buffer.from(h, "hex"));
-export const hexToNativeString = (h: string | undefined, c: ChainId) => {
-  try {
-    return !h
-      ? undefined
-      : c === CHAIN_ID_SOLANA
-      ? new PublicKey(hexToUint8Array(h)).toString()
-      : isEVMChain(c)
-      ? hexZeroPad(hexValue(hexToUint8Array(h)), 20)
-      : c === CHAIN_ID_TERRA
-      ? isHexNativeTerra(h)
-        ? nativeTerraHexToDenom(h)
-        : humanAddress(hexToUint8Array(h.substr(24))) // terra expects 20 bytes, not 32
-      : h;
-  } catch (e) {}
-  return undefined;
+
+/**
+ *
+ * Convert an address in a wormhole's 32-byte array representation into a chain's
+ * native string representation.
+ */
+export const uint8ArrayToNative = (
+  a: Uint8Array,
+  chainId: ChainId
+): string => {
+  if (isEVMChain(chainId)) {
+    return hexZeroPad(a, 20)
```

Comment:
Is it safe to have removed the hex value here?

Reply:
No, it's not... looks like `hexZeroPad` expects the string to start with `0x`...

Reply:
oh, except that we now pass in a `Uint8Array` which it handles differently - so this is fine. But I'll add tests